### PR TITLE
Added support keyboard Epomaker TH66

### PR DESCRIPTION
**Pull request by HorrorTroll**
## Description

Added support keyboard Epomaker TH66

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/epomaker/th66/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    [_FN_MAC] = LAYOUT(
+                KC_GRV,      KC_BRID, KC_BRIU, KC_MCTL, KC_SIRI, RGB_VAD, RGB_VAI, KC_MPRV, KC_MPLY, KC_MNXT, KC_MUTE, KC_VOLD, KC_VOLU, RGB_TOG,          QK_BOOT,
+                NK_TOGG,     _______, _______, _______, _______, _______, _______, _______, _______, _______, RGB_HUI, _______, _______, RGB_MOD,
+                _______,     _______, _______, _______, _______, _______, _______, _______, KC_HOME, KC_END,  _______, _______,          _______, _______,
+                _______,              _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, RGB_VAI, KC_DEL,
+                LCA(EE_CLR), GUI_TOG, WIN_MOD,                            _______,                   _______, _______,          _______, RGB_SPD, RGB_VAD, RGB_SPI
+            ),
+};
+
+void matrix_scan_user(void) {
```
Comment (drashna) [inline keyboards/epomaker/th66/keymaps/default/keymap.c:128]:
```suggestion
void housekeeping_task_user(void) {
```

File: keyboards/epomaker/th66/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    [_FN_MAC] = LAYOUT(
+                KC_GRV,      KC_BRID, KC_BRIU, KC_MCTL, KC_SIRI, RGB_VAD, RGB_VAI, KC_MPRV, KC_MPLY, KC_MNXT, KC_MUTE, KC_VOLD, KC_VOLU, RGB_TOG,          QK_BOOT,
+                NK_TOGG,     _______, _______, _______, _______, _______, _______, _______, _______, _______, RGB_HUI, _______, _______, RGB_MOD,
+                _______,     _______, _______, _______, _______, _______, _______, _______, KC_HOME, KC_END,  _______, _______,          _______, _______,
+                _______,              _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, RGB_VAI, KC_DEL,
+                LCA(EE_CLR), GUI_TOG, WIN_MOD,                            _______,                   _______, _______,          _______, RGB_SPD, RGB_VAD, RGB_SPI
+            ),
+};
+
+void matrix_scan_user(void) {
```
Comment (drashna) [inline keyboards/epomaker/th66/keymaps/via/keymap.c:128]:
```suggestion
void housekeeping_task_user.(void) {
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/epomaker/th66/keymaps/default/keymap.c (hunk 1 above)
Comment (HorrorTroll) [reply keyboards/epomaker/th66/keymaps/default/keymap.c]: done, thanks

**Review by drashna — APPROVED** [review]

File: keyboards/epomaker/th66/th66.c [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "th66.h"
+
+#ifdef RGB_MATRIX_ENABLE
+const is31_led PROGMEM g_is31_leds[DRIVER_LED_TOTAL] = {
```
Comment (adophoxia) [inline keyboards/epomaker/th66/th66.c:21]:
```suggestion
const is31_led PROGMEM g_is31_leds[RGB_MATRIX_LED_COUNT] = {
```

File: keyboards/epomaker/th66/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            rgb_matrix_set_color(31, 255, 0, 0);
+        }
+    } else {
+        if (host_keyboard_led_state().caps_lock) {
+            rgb_matrix_set_color(31, 255, 0, 0);
+        } else {
+            rgb_matrix_set_color(31, 0, 0, 0);
+        }
+    }
+}
```
Comment (filterpaper) [inline keyboards/epomaker/th66/keymaps/default/keymap.c:206]:
```suggestion
 bool rgb_matrix_indicators_user(void) {

    if ((rgb_matrix_get_flags() & LED_FLAG_KEYLIGHT)) {
        if (host_keyboard_led_state().caps_lock) {
            rgb_matrix_set_color(31, 255, 0, 0);
        }
    } else {
        if (host_keyboard_led_state().caps_lock) {
            rgb_matrix_set_color(31, 255, 0, 0);
        } else {
            rgb_matrix_set_color(31, 0, 0, 0);
        }
    }
    return false;
}
```

File: keyboards/epomaker/th66/keymaps/via/keymap.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            rgb_matrix_set_color(31, 255, 0, 0);
+        }
+    } else {
+        if (host_keyboard_led_state().caps_lock) {
+            rgb_matrix_set_color(31, 255, 0, 0);
+        } else {
+            rgb_matrix_set_color(31, 0, 0, 0);
+        }
+    }
+}
```
Comment (filterpaper) [inline keyboards/epomaker/th66/keymaps/via/keymap.c:206]:
```suggestion
 bool rgb_matrix_indicators_user(void) {

    if ((rgb_matrix_get_flags() & LED_FLAG_KEYLIGHT)) {
        if (host_keyboard_led_state().caps_lock) {
            rgb_matrix_set_color(31, 255, 0, 0);
        }
    } else {
        if (host_keyboard_led_state().caps_lock) {
            rgb_matrix_set_color(31, 255, 0, 0);
        } else {
            rgb_matrix_set_color(31, 0, 0, 0);
        }
    }
    return false;
}
```

**Review by filterpaper — CHANGES_REQUESTED** [review]

File: keyboards/epomaker/th66/keymaps/via/keymap.c (hunk 5 above)
Comment (adophoxia) [inline keyboards/epomaker/th66/keymaps/via/keymap.c:206]:
```suggestion
 bool rgb_matrix_indicators_user(void) {
    if ((rgb_matrix_get_flags() & LED_FLAG_KEYLIGHT)) {
        if (host_keyboard_led_state().caps_lock) {
            rgb_matrix_set_color(31, 255, 0, 0);
        }
    } else {
        if (host_keyboard_led_state().caps_lock) {
            rgb_matrix_set_color(31, 255, 0, 0);
        } else {
            rgb_matrix_set_color(31, 0, 0, 0);
        }
    }
    return false;
}
```
Refer to https://github.com/qmk/qmk_firmware/pull/18448

**Comment by TheMohawkNinja** [comment]
Not sure if some of the previous changes made to the original upload of the TH66 firmware resolves this, but I noticed my TH66 had a similar issue to the one presented in /issues/6726 (more specifically, turning the PC off without disconnecting the keyboard would result in apparent corruption of the firmware after some time had passed, necessitating re-flashing). The stated solution of adding `#define NO_USB_STARTUP_CHECK` to `config.h` also resolved my issue.

**Comment by HorrorTroll** [comment]
Done, updated the entire code with new info.json thing.

File: keyboards/epomaker/th66/config.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#ifdef RGB_MATRIX_ENABLE
+    #define RGB_MATRIX_LED_COUNT (DRIVER_1_LED_TOTAL + DRIVER_2_LED_TOTAL)
+    #define RGB_MATRIX_MAXIMUM_BRIGHTNESS 200
+    #define RGB_MATRIX_STARTUP_VAL RGB_MATRIX_MAXIMUM_BRIGHTNESS
```
Comment (fauxpark) [inline keyboards/epomaker/th66/config.h:23]:
```suggestion
    #define RGB_MATRIX_DEFAULT_VAL RGB_MATRIX_MAXIMUM_BRIGHTNESS
```

File: keyboards/epomaker/th66/info.json [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    "keyboard_name": "TH66",
+    "manufacturer": "Epomaker",
+    "url": "https://www.aliexpress.com/item/1005004353520898.html",
+    "maintainer": "Epomaker",
+    "usb": {
+        "vid": "0x586A",
+        "pid": "0x0060",
+        "device_version": "0.0.1"
+    },
+    "debounce": 5,
```
Comment (fauxpark) [inline keyboards/epomaker/th66/info.json:11]:
```suggestion
```

File: keyboards/epomaker/th66/info.json [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            {"matrix": [5, 0], "x": 2, "y": 64, "flags": 1},
+            {"matrix": [5, 1], "x": 20, "y": 64, "flags": 1},
+            {"matrix": [3, 14], "x": 224, "y": 20, "flags": 4},
+            {"matrix": [3, 15], "x": 224, "y": 36, "flags": 4}
+        ]
+    },
+    "layouts": {
+        "LAYOUT": {
+            "layout": [
+                {"matrix": [0, 0], "w": 1, "x": 0, "y": 0},
```
Comment (fauxpark) [inline keyboards/epomaker/th66/info.json:177]:
```suggestion
                {"matrix": [0, 0], "x": 0, "y": 0},
```
etc...

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/epomaker/th66/info.json (hunk 8 above)
Comment (HorrorTroll) [reply keyboards/epomaker/th66/info.json]: Thanks for reviewing, i've pushed new commit to resolved those.

File: keyboards/epomaker/th66/th66.c [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+                tap_code(KC_TRNS);
+            } else {
+                tap_code(KC_TRNS);
+            }
+        }
+    }
+    return true;
+}
+#endif
+
```
Comment (waffle87) [inline keyboards/epomaker/th66/th66.c:55]: This is/will be the core default behaviour.
```suggestion
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**